src/compiler/checker.ts

```diff
@@ -5603,18 +5603,29 @@ namespace ts {
             return compareTypes(getTypeOfSymbol(sourceProp), getTypeOfSymbol(targetProp));
         }
 
+        // A source signature matches a target signature if the two signatures have the same number of required,
+        // optional, and rest parameters.
+        function isMatchingSignature(source: Signature, target: Signature) {
+            return source.parameters.length === target.parameters.length &&
+                source.minArgumentCount === target.minArgumentCount &&
+                source.hasRestParameter === target.hasRestParameter;
+        }
+
+        // A source signature partially matches a target signature if the target signature has no fewer required
+        // parameters and no more overall parameters than the source signature (where a signature with a rest
+        // parameter is always considered to have more overall parameters than one without).
+        function isPartiallyMatchingSignature(source: Signature, target: Signature) {

```

Comment:
I don't exactly see why it doesn't make sense. You'd return `true` if ordinary matched-ness worked, and then if `allowPartialMatches` was given in, you'd just perform the extra check. Otherwise you'd return false.